### PR TITLE
chef sample app

### PR DESCRIPTION
**Pull request by rochaferraz**
#### Description
The purpose of the chef app is to to increase the coverage of device types in
Matter.

It uses the shell app a starting point, adding the processing of ZAP files and
the support of a few targets under a unified build script: `chef.py`.

As it incorporates the processing of ZAP files as part of the build process, it
does not use `zzz_generated`, but rather places the auto-generated zap
artifacts under its `zap-generated` temporary folder.

All device types available (DM/IM .zap files) are found inside the `devices`
folder.

#### Change overview
New sample App under examples/chef

_https://github.com/project-chip/connectedhomeip/pull/16639 is similar and targets [interop_testing_2022_01_18](https://github.com/project-chip/connectedhomeip/tree/interop_testing_2022_01_18) branch_

### CODE REVIEW

File: examples/chef/esp32/main/main.cpp [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+#if CONFIG_ENABLE_CHIP_SHELL
+    int ret = Engine::Root().Init();
+    VerifyOrDie(ret == 0);
+#endif /* CONFIG_ENABLE_CHIP_SHELL */
+
+    // Network connectivity
+    // Note to integration: StartWiFiManagement does not exist on ESP32
+
+    // Start IM server
+    chip::Server::GetInstance().Init();
```
Comment (dnicoara) [inline examples/chef/esp32/main/main.cpp:177]: Needs to be done on the chip thread ... something like `chip::DeviceLayer::PlatformMgr().ScheduleWork(...)`

File: examples/chef/esp32/main/main.cpp [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    // Device Attestation & Onboarding codes
+    chip::Credentials::SetDeviceAttestationCredentialsProvider(chip::Credentials::Examples::GetExampleDACProvider());
+    chip::DeviceLayer::ConfigurationMgr().LogDeviceConfig();
+    PrintOnboardingCodes(chip::RendezvousInformationFlag(chip::RendezvousInformationFlag::kBLE));
+
+#if CONFIG_HAVE_DISPLAY
+    printQRCode();
+#endif // CONFIG_HAVE_DISPLAY
+
+    if (chip::Server::GetInstance().GetCommissioningWindowManager().OpenBasicCommissioningWindow() != CHIP_NO_ERROR)
```
Comment (dnicoara) [inline examples/chef/esp32/main/main.cpp:188]: This also needs to be on the chip thread.

File: examples/chef/esp32/main/main.cpp [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    // Initialize the screen manager
+    ScreenManager::Init();
+
+    ESP_LOGI(TAG, "Opening QR code screen");
+    ESP_LOGI(TAG, "QR CODE Text: '%s'", qrCodeText.c_str());
+    ScreenManager::PushScreen(chip::Platform::New<QRCodeScreen>(qrCodeText));
+}
+#endif // CONFIG_HAVE_DISPLAY
+
+void InitServer(intptr_t context)
```
Comment (dnicoara) [inline examples/chef/esp32/main/main.cpp:161]: The context isn't used anywhere. Might as well remove it.

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
Fast tracking testing changes.

@rochaferraz Can you please re-generate zap?

File: examples/chef/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+folder.
+
+## Building a Sample Application
+
+Run `chef.sh -h` to see the available commands
+
+## Building your first sample
+
+1. Make sure you have the toolchain installed for your desired target
+2. Update your SoC SDK paths on `chef_config.sh` and flip the `configured`
```
Comment (mspang) [inline examples/chef/README.md:23]: Outdated path?

File: examples/chef/chef_config.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import os
+#
+# USER DEFINED INFORMATION
```
Comment (mspang) [inline examples/chef/chef_config.py:3]: Why are we doing configuration by editing an existing file? This will dirty the tree.

**Comment by mspang** [comment]
Should it merged into shell/ ?

File: examples/chef/chef_config.py (hunk 5 above)
Comment (rochaferraz) [reply examples/chef/chef_config.py:2]: fixed, it wasn't needed

File: examples/chef/chef_config.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+#
+configured = False
+#configured = True
+
+# NRFConnect
+
+nrfconnectFolder = "/opt/nordic/ncs/v1.9.1"
+nrfConnectSerialDevice = "/dev/tty.usbmodem0*"  # Typical MacOS path
+# nrfConnectSerialDevice = "/dev/ttyACM0" # Typical Linux path
```
Comment (andy31415) [inline examples/chef/chef_config.py:11]: this is python - could we check `sys.platform` or something like that?

Comment (andy31415) [reply examples/chef/chef_config.py:11]: Can be done as a followup: I see that the code asks for `configured=True` to be done.

Generally I believe this has to be fixed: dirtying the tree should not be required to build/configure things. At most, we can ask for a git-ignore file for settings (e.g. sdkconfig does this) and have python scripts as helpers here.

Also folders for NRFConnect and similar can be defaulted from OS environment variables, like our build scripts do.

**Review by andy31415 — APPROVED** [review]

**Pull request merged**